### PR TITLE
Add Deepstack/CodeProject-AI.Server detector plugin

### PR DESCRIPTION
**Pull request by skrashevich**
example configuration:
```
deepstack:
  api_url: http://192.168.88.50:5500/v1/vision/detection
  type: deepstack
  api_timeout: 0.1
```
docker container for test: [skrashevich/frigate:deepstack](https://hub.docker.com/layers/skrashevich/frigate/deepstack/images/sha256-7a536037f446d1606ec8beb63ecaa0868bd512218cedb7efc533f39929ddcda3?context=explore)

### CODE REVIEW

**Comment by NickM-27** [comment]
What kinds of inference times are you seeing?

**Comment by skrashevich** [comment]
6 cameras, each on 10fps. frigate and ai.server on different physical machines:
telegram-cloud-photo-size-2-5330101685202307428-y

**Comment by NickM-27** [comment]
thanks, and what is the deepstack server using to run the inferences (CPU, gpu, etc?)

**Comment by skrashevich** [comment]
CodeProject Ai.Server 2.1.1, YOLOv5 6.2 Large model, half-precision. GPU NVIDIA GeForce GTX 1070

**Comment by hawkeye217** [comment]
I've been seeing some hype around the interwebs lately about CodeProject AI, so I wanted to experiment with some of the custom models they've trained (["ipcam-dark"](https://www.codeproject.com/ai/docs/api/api_reference.html#custom-object-detector) specifically). I tested this PR by quickly spinning up CodeProject AI via docker-compose:

```
  codeprojectai:
    container_name: codeprojectai
    image: codeproject/ai-server:2.1.3
    ports:
      - 32168:32168
    restart: always
    volumes:
      - /home/myuser/docker/codeprojectai/etc:/etc/codeproject/ai
      - /home/myuser/docker/codeprojectai/modules:/app/modules
```

config.yml:

```
detectors:
  deepstack:
    api_url: http://192.168.1.4:32168/v1/vision/custom/ipcam-dark
    type: deepstack
    api_timeout: 1
```

The particular ipcam-dark model returns a label with a capital first letter, so that threw off Frigate. A quick hack on line 78 of this PR fixed it:

```
int(self.get_label_index(detection["label"].lower()))
```

Additionally, the default YOLOv5 model in CodeProject seemed to return labels that weren't in `/labelmap.txt` - I didn't dig any deeper, but Frigate crashed whenever that happened.

Overall though, I love the flexibility to have various detectors as plugins.

**Comment by skrashevich** [comment]
> Additionally, the default YOLOv5 model in CodeProject seemed to return labels that weren't in `/labelmap.txt` - I didn't dig any deeper, but Frigate crashed whenever that happened.
Please show the logs from the moment the frigate crashes

**Comment by hawkeye217** [comment]
> > Additionally, the default YOLOv5 model in CodeProject seemed to return labels that weren't in `/labelmap.txt` - I didn't dig any deeper, but Frigate crashed whenever that happened.
> 
> Please show the logs from the moment the frigate crashes

```
2023-04-23 08:17:22.489122981    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2023-04-23 08:17:22.489124149      self.run()
2023-04-23 08:17:22.489125520    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2023-04-23 08:17:22.489126709      self._target(*self._args, **self._kwargs)
2023-04-23 08:17:22.489128057    File "/opt/frigate/frigate/object_detection.py", line 121, in run_detector
2023-04-23 08:17:22.489129311      detections = object_detector.detect_raw(input_frame)
2023-04-23 08:17:22.489130588    File "/opt/frigate/frigate/object_detection.py", line 71, in detect_raw
2023-04-23 08:17:22.489131841      return self.detect_api.detect_raw(tensor_input=tensor_input)
2023-04-23 08:17:22.489133171    File "/opt/frigate/frigate/detectors/plugins/deepstack.py", line 78, in detect_raw
2023-04-23 08:17:22.489134415      int(self.get_label_index(detection["label"])),
2023-04-23 08:17:22.489135776  TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

**Comment by hawkeye217** [comment]
So the issue here specifically has to do with the label "truck".

In `/labelmap.txt`, `7 truck` has been renamed to `7 car` - this was a decision by the devs a while ago as the default model has a hard time distinguishing between car and truck.

When the CodeProject model returns "truck", this is what causes Frigate to crash.

**Comment by skrashevich** [comment]
Probably fixed

**Comment by NickM-27** [comment]
Please use black to reformat so check can pass

**Comment by skrashevich** [comment]
Done
+ Some fixes to use build-in label map instead of custom loader

**Comment by hawkeye217** [comment]
> Hmm, then maybe I could spin up a second Frigate with deepstack as the detector and use it just for detecting objects / logos? Either way, this is an exciting enhancement. Thanks!

That's probably what you'd have to do at this point, and you'd need to also [customize your labelmap](https://docs.frigate.video/configuration/advanced/#model) based on your model.

**Comment by skrashevich** [comment]
> I have a Coral USB that I'm using with Frigate. If deepstack were added as a detector and if I still wanted to use the coral for other objects, would I just set up a second camera with the same restream that uses only the deepstack detector? I wouldn't be recording, just detecting objects and passing along the values back into HA for consumption / sending alerts. I hate porch pirates.

See #6258

**Comment by blakeblackshear** [comment]
This just needs docs updates before it can be merged.

**Comment by NickM-27** [comment]
I'd say the main section of https://docs.frigate.video/configuration/detectors#openvino-detector would be a good template. Basically you want to have:

```
# <title>

<brief description of CodeProject.AI and the devices it can be run on. I think it would be important for you to mention it is over the network so inference times won't be as good as a native detector.>

## Setup

<describe where to get codeproject ai, I think full setup instructions are outside the scope of the frigate docs>

<show the frigate config file changes that need to be made>

<describe how to verify it is working>
```

**Comment by skrashevich** [comment]
Here's the question, describe deepstack or codeproject ai.server in Setup section?

**Comment by NickM-27** [comment]
> Here's the question, describe deepstack or codeproject ai.server in Setup section?

Given that this is directly interfacing with deepstack, it would be deepstack in that case

**Comment by NickM-27** [comment]
> > Given that this is directly interfacing with deepstack, it would be deepstack in that case
> 
> but deepstack and ai.server implements the identical API (ai.server can be used as drop-in replacement of deepstack). Deepstack was the first, but not maintained now, unlike ai.server

I see, that makes sense. In that case I would state exactly that, and mainly focus on codeproject ai

**Comment by skrashevich** [comment]
🤷

**Comment by sstratoti** [comment]
Re: deepstack being maintained...

https://github.com/johnolafenwa/DeepStack/issues/182

He posted this 3 months ago. No new commits, but it's not entirely dead yet.

**Comment by skrashevich** [comment]
> He posted this 3 months ago. No new commits, but it's not entirely dead yet.

but latest release was more than year ago...
anyway, punks not dead! :)

File: docs/docs/configuration/detectors.md [hunk 1]
```diff
@@ -256,3 +256,24 @@ model:
   width: 416
   height: 416
 ```
+
+
+## Deepstack / CodeProject AI.Server Detector
+The Deepstack/CodeProject AI.Server detector plugin for Frigate allows you to integrate Deeppstack or CodeProject.AI object detection capabilities into Frigate video surveillance system. CodeProject.AI and DeepStack is an open-source AI platforms that can be run on various devices, such as Raspberry Pi, Nvidia Jetson, and other compatible hardware. It is important to note that the integration is performed over the network, so the inference times may not be as fast as native Frigate detectors, but it still provides an efficient and reliable solution for object detection and tracking.
```
Comment (NickM-27) [inline docs/docs/configuration/detectors.md:262]:
```suggestion

## Deepstack / CodeProject.AI Server Detector

The Deepstack / CodeProject.AI Server detector for Frigate allows you to integrate Deepstack and CodeProject.AI object detection capabilities into Frigate. CodeProject.AI and DeepStack are open-source AI platforms that can be run on various devices such as the Raspberry Pi, Nvidia Jetson, and other compatible hardware. It is important to note that the integration is performed over the network, so the inference times may not be as fast as native Frigate detectors, but it still provides an efficient and reliable solution for object detection and tracking.
```

File: docs/docs/configuration/detectors.md [hunk 2]
```diff
@@ -256,3 +256,24 @@ model:
...
+
+### Setup
+To get started with CodeProject.AI, visit their official website at [www.codeproject.com](https://www.codeproject.com/Articles/5322557/CodeProject-AI-Server-AI-the-easy-way) and follow the instructions to download and install the AI server on your preferred device. Detailed setup instructions for CodeProject.AI are outside the scope of the Frigate documentation.
+
+To integrate CodeProject.AI with Frigate, you'll need to make the following changes to your Frigate configuration file:
```
Comment (NickM-27) [inline docs/docs/configuration/detectors.md:267]:
```suggestion
### Setup

To get started with CodeProject.AI, visit their [official website](https://www.codeproject.com/Articles/5322557/CodeProject-AI-Server-AI-the-easy-way) to follow the instructions to download and install the AI server on your preferred device. Detailed setup instructions for CodeProject.AI are outside the scope of the Frigate documentation.

To integrate CodeProject.AI into Frigate, you'll need to make the following changes to your Frigate configuration file:
```

**Review by NickM-27 — CHANGES_REQUESTED** [review]

**Review by NickM-27 — APPROVED** [review]

**Comment by skrashevich** [comment]
Perhaps it makes sense to add information that the documentation on codeproject.com is dramatically outdated, and user need to use another port (5000 instead of 32168) to run in the docker container?

**Comment by NickM-27** [comment]
Interesting, that port worked for me when testing code project ai on my server on the latest version.

**Comment by hawkeye217** [comment]
> Perhaps it makes sense to add information that the documentation on codeproject.com is dramatically outdated, and user need to use another port (5000 instead of 32168) to run in the docker container?

I used those instructions and port 32168 when spinning up my container when testing this PR and didn't have any issues.

**Comment by skrashevich** [comment]
What versions of the container and host OS (windows, linux,mac)?

**Comment by NickM-27** [comment]
> What versions of the container and host OS (windows, linux,mac)?

Linux, pulling the latest `codeproject/ai-server:gpu`

**Comment by skrashevich** [comment]
> Linux, pulling the latest `codeproject/ai-server:gpu`
the same image on linux for me works on port 5000. can you check this port on you installation? :)

**Comment by hawkeye217** [comment]
> What versions of the container and host OS (windows, linux,mac)?

Similar to Nick, Linux, using `codeproject/ai-server`, non-gpu though.

My docker-compose was [here](https://github.com/blakeblackshear/frigate/pull/6143#issuecomment-1518474242)

**Comment by skrashevich** [comment]
It's a bug in AI.Server. in some cases, related to docker network type (macvlan of ipvlan) it may stops listen on 32168.
need to keep in mind...

**Comment by skrashevich** [comment]
> did you ever try out the `ipcam` models from Codeproject? just curious how good they were

They work worse than the main.

The best thing you can do: fetch `https://github.com/ultralytics/yolov5/releases/download/v7.0/yolov5x6.pt`, save to `/app/preinstalled-modules/ObjectDetectionYolo/custom-models/ipcam-yolov5x6` and use in this detector: `api_url: http://192.168.88.50:5500/v1/vision/custom/ipcam-yolov5x6`
Particularly good within PR #6429

**Comment by inspired27** [comment]
> > did you ever try out the `ipcam` models from Codeproject? just curious how good they were
> 
> They work worse than the main.
> 
> The best thing you can do: fetch `https://github.com/ultralytics/yolov5/releases/download/v7.0/yolov5x6.pt`, save to `/app/preinstalled-modules/ObjectDetectionYolo/custom-models/ipcam-yolov5x6` and use in this detector: `api_url: http://x.x.x.x:nnnn/v1/vision/custom/ipcam-yolov5x6` Even good within PR #6429

Hi,

I want to use the license-plate model in codeproject.ai with the Frigate deepstack detector, I have codeproject.ai server running and detector pointing to the license-plate model as below but it is not detecting. The labels returned I think are Dayplate and Nightplate but I do not know how to make this work with Frigate is there something else I need to update? I am running dev 0.13.0-0996883 in docker. I am new so some simple steps instructions would really help.

detectors:
  deepstack:
    api_url: http://192.168.1.xx:32168/v1/vision/custom/license-plate
    type: deepstack
    api_timeout: 1 # seconds

**Pull request merged**